dask/tests/test_tokenize.py

```diff
@@ -62,69 +64,173 @@ def test_normalize_function():
 
 
+def _clear_function_cache():
+    from dask.base import function_cache, function_cache_lock
+
+    with function_cache_lock:
+        function_cache.clear()
+
+
+def check_tokenize(*args, idempotent=True, deterministic=None, copy=None, **kwargs):
+    """Wrapper around tokenize
+
+    Parameters
+    ----------
+    args, kwargs: passed to tokenize
+    idempotent: True or False
+        If True, expect tokenize() called on the same object twice to produce the same
+        result. If False, expect different results. Default: True
+    deterministic: True, False, or "maybe"
+        If True, expect tokenize() called on two identical copies of an object to
+        produce the same result. If False, expect different results. If "maybe", expect
+        nothing. Default: same as idempotent
+    copy: callable or False
+        f(T)->T that deep-copies the object. Default: distributed serialize
+    """
+    if deterministic is None:
+        deterministic = idempotent
+
+    if copy is None:
+        copy = lambda x: cloudpickle.loads(cloudpickle.dumps(x))
+
+    ensure_deterministic = deterministic is True  # not maybe
+    with dask.config.set({"tokenize.ensure-deterministic": ensure_deterministic}):
+        before = tokenize(*args, **kwargs)
+
+        # Test idempotency (the same object tokenizes to the same value)
+        _clear_function_cache()
+        after = tokenize(*args, **kwargs)
+
+        assert (before == after) is idempotent
+
+        # Test same-interpreter determinism (two identical objects tokenize to
+        # the same value as long as you do it on the same interpreter)
+        if copy:
+            args2, kwargs2 = copy((args, kwargs))
+            _clear_function_cache()
+            after = tokenize(*args2, **kwargs2)
+
+            if deterministic != "maybe":
+                assert (before == after) is deterministic
+
+        # Skip: different interpreter determinism
+
+    return before
+
+
+def test_check_tokenize():
+    # idempotent and deterministic
+    check_tokenize(123)
+
+    # idempotent, but not deterministic
+    class A:
+        def __init__(self):
+            self.tok = random.random()
+
+        def __reduce_ex__(self, protocol):
+            return A, ()
+
+        def __dask_tokenize__(self):
+            return self.tok
+
+    a = A()
+    with pytest.raises(AssertionError):
+        check_tokenize(a)
+    check_tokenize(a, deterministic=False)
+    check_tokenize(a, deterministic="maybe")
+
+    # Not idempotent
+    class B:
+        def __dask_tokenize__(self):
+            return random.random()
+
+    b = B()
+    check_tokenize(b, idempotent=False)
+    with pytest.raises(AssertionError):
+        check_tokenize(b)
+    with pytest.raises(AssertionError):
+        check_tokenize(b, deterministic=False)
+
+
 def test_tokenize():
     a = (1, 2, 3)
-    assert isinstance(tokenize(a), (str, bytes))
+    assert isinstance(check_tokenize(a), (str, bytes))
```

Comment:
```suggestion
    assert isinstance(tokenize(a), (str, bytes))
```
since we now test `check_tokenize` in its own test.